cmd/manager/cleanup.go

```diff
@@ -0,0 +1,178 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	olm "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	marketplace "github.com/operator-framework/operator-marketplace/pkg/apis/operators/v1"
+	"github.com/sirupsen/logrus"
+	k8s_errors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// NewCleanUpResources returns a CleanUpCrew that updates existing
+// subscriptions linked to old CSCs and links them to the datastore CSCs
+// in the openshift-marketplace namespace, and then deletes those CSCs
+// along with existing datastore CSCs.
+func NewCleanUpResources(logger *logrus.Entry, client client.Client) CleanUpCrew {
+	return CleanUpCrew{
+		logger: logger,
+		client: client,
+	}
+}
+
+type CleanUpCrew struct {
+	logger *logrus.Entry
+	client client.Client
+}
+
+func (c *CleanUpCrew) CleanupOldResource() error {
+
+	c.logger.Printf("Updating existing subscriptions.")
+	datastoreCscNames, err := updateSubsciptions(c.client)
+	if err != nil {
+		return err
+	}
+
+	c.logger.Printf("Deleting datastore CatalogSourceConfigs.")
+	err = deleteDatastoreCscs(datastoreCscNames, c.client)
+	if err != nil {
+		return err
+	}
+
+	c.logger.Printf("Deleting old CatalogSourceConfigs created duing operator installations.")
+	err = deleteOldCscs(c.client)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func updateSubsciptions(kubeClient client.Client) ([]string, error) {
+	var cscNames []string
+	options := &client.ListOptions{}
+	options.SetLabelSelector(fmt.Sprintf("csc-owner-name"))
+
+	subscriptions := &olm.SubscriptionList{}
+	// Get the list of existing subscriptions that have the label "csc-owner-name"
+	err := kubeClient.List(context.TODO(), options, subscriptions)
+	if err != nil {
+		return nil, err
+	}
+	for _, instance := range subscriptions.Items {
+		cscNames = append(cscNames, instance.GetLabels()["csc-owner-name"])
+		// try to guess the datastore catalogsource from the subscription
+		datastoreCs, err := findCatalogSource(&instance, kubeClient)
+		if k8s_errors.IsNotFound(err) {
+			// guess the catalogsource from the opsrc that has the package
+			datastoreCs, err = findCsFromOpsrc(&instance, kubeClient)
+			if err != nil {
+				return nil, err
+			}
+		}
+		// point the subscription to the datastore CatalogSource
+		instance.SetCatalogSource(datastoreCs.GetName())
+		instance.SetCatalogSourceNamespace("openshift-marketplace")
```

Comment:
I see where you are hard-coding the namespace. This is not required. You should be able to infer this. This applies to all instances where you are doing this.